components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    ops::Deref,
+    path::{Path, PathBuf},
+};
+
+use once_cell::sync::OnceCell;
+use remote_settings::{self, Attachment, GetItemsOptions, RemoteSettingsConfig, SortOrder};
+use serde_derive::*;
+
+use crate::{
+    db::{ConnectionType, SuggestDb, LAST_FETCH_META_KEY},
+    RemoteRecordId, RemoteSuggestion, Result, SuggestApiResult, Suggestion, SuggestionQuery,
+};
+
+const RS_COLLECTION: &str = "quicksuggest";
+
+/// The store is the entry point to the Suggest component. It incrementally
+/// fetches suggestions from the Remote Settings service, stores them in a local
+/// database, and returns them in response to user queries.
+///
+/// Your application should create a single store, and manage it as a singleton.
+/// The store is thread-safe, and supports concurrent fetches and ingests. We
+/// expect that your application will call `query` to show suggestions as the
+/// user types into the address bar, and periodically call `ingest` in the
+/// background to update the database with new suggestions from Remote Settings.
+///
+/// The store keeps track of the state needed to support incremental ingestion,
+/// but doesn't schedule the ingestion work itself, or decide how many
+/// suggestions to ingest at once. This is for two reasons:
+///
+/// 1. The primitives for scheduling background work vary between platforms. You
+///    might use an idle timer on Desktop, `WorkManager` on Android, or
+///    `BGTaskScheduler` on iOS.
+/// 2. Ingestion limits can change, depending on the platform and the needs of
+///    the application. A mobile device on a metered connection might want to
+///    request a small subset of the Suggest data and fetch the rest later,
+///    while a desktop on a fast link might request the entire dataset on first
+///    launch.
```

Comment:
Thanks, I found it really helpful to write this when sketching out the architecture, too!

Also, based on our other conversation, this might be a good place to mention `interrupt` 😁👍🏼